### ISSUE TITLE
twisted.web.distrib non-functional

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2681]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2681|
|Type|defect|
|Created|2007-05-28 20:33:19Z|
|Branch|https://github.com/twisted/twisted/tree/personal-servers-2681|

twisted/web/tap.py tries to instantiate a pb.BrokerFactory in order to set up a _distrib_ or _personal_ server.  This fails, as pb.BrokerFactory was removed when twisted.internet.application was removed.

Attachments:

* [web_tap_pb_broker.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3e1/3e11fda4b1c58ea4bc6b608c03913c5b3d7fbde7/abf1f91ffbaee9a49c1efda2f88a15898e2a3674.patch) (588 bytes) - added by straggler77 on 2007-07-02 05:10:52Z - 
* [unit_test_web_tap.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3e1/3e11fda4b1c58ea4bc6b608c03913c5b3d7fbde7/7257192c924cc0803653046f5fcd23c891992bcf.py) (356 bytes) - added by hagna on 2007-11-19 22:56:22Z - one unit test that tests the bug the patch fixes

<details><summary>Searchable metadata</summary>

```
trac-id__2681 2681
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_personal_servers_2681 branches/personal-servers-2681
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1180384399000000 1180384399000000
changetime__1202849111000000 1202849111000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="straggler77's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|straggler77 commented|
|-|-|

The attached patch appears to fix this issue.

The issue can be recreated by issuing the following command:
mktap web --personal

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to straggler77|
|-|-|

Thanks.  Can you write a unit test, too?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to hagna|
|-|-|

Good start.  I can see how you would get here - the function call in that test method will raise an exception until this bug is removed, so fixing it will make the unit test pass.  Deleting all of the code from the if suite which handles --personal will also make the test pass, though.  There should be an assertion in the test method.  You might want to factor the distrib setup code into a separate method which makeService will call and the unit test can call directly so that it gets a distrib resource back directly and can make assertions about it (for example, assertions about the values of its attributes) instead of having to deal with the larger, more opaque object returned by makeService.

**Comment by twisted-trac** [comment]
|<img alt="hagna's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|hagna commented|
|-|-|

what distrib resource and what distrib setup code?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [22518]) Branching to 'personal-servers-2681'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @exarkun|
|-|-|

 * trailing whitespaces in tap.py
 * the docstring of tap.py use the 'I am' style, it would be cool if could change this.
 * makePersonalServerFactory should have a docstring
 * please use assertIsInstance in test_makePersonalServerFactory

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner|
|-|-|

Okay, did those things.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @exarkun|
|-|-|

Thanks, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [22538]) Merge personal-servers-2681

Author: exarkun
Reviewer: therve
Fixes [#2681](https://github.com/twisted/twisted/issues/2681)

Factor creation of the distrib/personal server factory into a separate function
which can easily be unit tested, add a test for it, and fix it so that it uses
a PBServerFactory (a class which exists) instead of BrokerFactory (a class which
does not exist).

This fixes "twistd web --personal".

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
